File: python/policy_guard.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass
class PolicyError(Exception):
    message: str
    code: str


class PolicyGuard:
    def __init__(self, allowlist: set[str], high_risk_tools: set[str]) -> None:
        self._allowlist = allowlist
        self._high_risk_tools = high_risk_tools

    def assert_device_allowed(self, device_id: str) -> None:
        if len(self._allowlist) == 0:
            return

        if device_id not in self._allowlist:
            raise PolicyError(message=f"Device is not allowed: {device_id}", code="ERR_DEVICE_NOT_ALLOWED")

    def assert_tool_risk(self, tool_name: str, args: dict[str, object]) -> None:
        if tool_name not in self._high_risk_tools:
            return

        confirm = args.get("confirm")
        reason = args.get("reason")
        if confirm is not True or not isinstance(reason, str) or len(reason.strip()) == 0:
            raise PolicyError(
                message="High-risk tool requires confirm=true and non-empty reason",
                code="ERR_CONFIRM_REQUIRED",
            )

File: src/policyGuard.ts
import { AppError } from "./errors.js";

export interface PolicyGuardOptions {
  allowlist: string[];
  highRiskTools?: string[];
}

export interface ToolRiskInput {
  toolName: string;
  args: Record<string, unknown>;
}

export class PolicyGuard {
  private readonly allowlist: Set<string>;
  private readonly highRiskTools: Set<string>;

  constructor(options: PolicyGuardOptions) {
    this.allowlist = new Set(options.allowlist);
    this.highRiskTools = new Set(options.highRiskTools ?? []);
  }

  assertDeviceAllowed(deviceId: string): void {
    if (this.allowlist.size === 0) {
      return;
    }

    if (!this.allowlist.has(deviceId)) {
      throw new AppError({
        message: `ERR_DEVICE_NOT_ALLOWED: Device is not allowed: ${deviceId}`,
        code: "ERR_DEVICE_NOT_ALLOWED",
        category: "business",
        traceId: "policy-check"
      });
    }
  }

  assertToolRisk(input: ToolRiskInput): void {
    if (!this.highRiskTools.has(input.toolName)) {
      return;
    }

    const confirm = input.args.confirm;
    const reason = input.args.reason;
    const hasReason = typeof reason === "string" && reason.trim().length > 0;

    if (confirm !== true || !hasReason) {
      throw new AppError({
        message: "ERR_CONFIRM_REQUIRED: High-risk tool requires confirm=true and non-empty reason",
        code: "ERR_CONFIRM_REQUIRED",
        category: "validation",
        traceId: "policy-check"
      });
    }
  }
}
